File: truss-compiler/src/test/java/truss/compiler/FixtureBase.java
package truss.compiler;

import truss.compiler.parser.TrussLexer;
import truss.compiler.parser.TrussParser;
import truss.compiler.syntax.CompilationUnitSyntax;
import org.antlr.runtime.ANTLRStringStream;
import org.antlr.runtime.CommonTokenStream;
import org.apache.commons.io.IOUtils;

import java.io.InputStream;

public abstract class FixtureBase implements TestCaseCallback {
    @Override
    public abstract void test(String name, String fileName, String code, String expected) throws Exception;

    protected void test(String resourceName) throws Exception {
        try (InputStream is = getClass().getResourceAsStream(resourceName)) {
            TestCase testCase = null;

            for (String line : IOUtils.readLines(is)) {
                if (line.equals("<<<")) {
                    testCase.beginExpected();
                } else if (line.startsWith(">>>")) {
                    if (testCase != null) {
                        testCase.test(this);
                    }
                    testCase = new TestCase(line.substring(3).trim(), resourceName);
                } else if (testCase != null) {
                    testCase.append(line);
                }
            }

            if (testCase != null) {
                testCase.test(this);
            }
        }
    }

    protected CompilationUnitSyntax parse(String name, String code) throws Exception {
        TrussLexer lexer = new TrussLexer(new ANTLRStringStream(code));
        lexer.setFileName(name);
        TrussParser parser = new TrussParser(new CommonTokenStream(lexer));
        parser.setFileName(name);

        CompilationUnitSyntax compilationUnit;

        try {
            compilationUnit = parser.compilationUnit();
        } finally {
            if (MessageCollectionScope.getCurrent().hasMessages()) {
                compilationUnit = null;
            }
        }

        return compilationUnit;
    }

    protected InputStream getSystemResource() {
        return FixtureBase.class.getResourceAsStream("System.dtl");
    }
}

File: Truss.Compiler.Test/FixtureBase.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using Antlr.Runtime;
using NUnit.Framework;
using Truss.Compiler.Parser;
using Truss.Compiler.Syntax;

namespace Truss.Compiler.Test {
    public abstract class FixtureBase : ITestCaseCallback {
        static FixtureBase() {
            if (!Debugger.IsAttached) {
                Debug.Listeners.Clear();
                Debug.Listeners.Add(new ThrowingTraceListener());
            }
        }

        public abstract void Test(String name, String fileName, String code, String expected);

        protected void Test(String resourceName) {
            using (var stream = GetType().Assembly.GetManifestResourceStream(GetType().Namespace + "." + resourceName))
            using (var reader = new StreamReader(stream)) {
                TestCase testCase = null;
                string line;

                while ((line = reader.ReadLine()) != null) {
                    if (line == "<<<") {
                        testCase.BeginExpected();
                    } else if (line.StartsWith(">>>")) {
                        if (testCase != null) {
                            testCase.Test(this);
                        }
                        testCase = new TestCase(line.Substring(3).Trim(), resourceName);
                    } else if (testCase != null) {
                        testCase.Append(line);
                    }
                }

                if (testCase != null) {
                    testCase.Test(this);
                }
            }
        }

        protected CompilationUnitSyntax Parse(ErrorList errors, string name, string code) {
            var lexer = new TrussLexer(new ANTLRStringStream(code));
            lexer.FileName = name;
            lexer.Errors = errors;
            var parser = new TrussParser(new CommonTokenStream(lexer));
            parser.FileName = name;
            parser.Errors = errors;

            CompilationUnitSyntax compilationUnit;

            try {
                compilationUnit = parser.ParseCompilationUnit();
            } catch (Exception e) {
                errors.Add(Error.InternalError, name, e.Message);

                compilationUnit = null;
            } finally {
                if (errors.HasMessages) {
                    compilationUnit = null;
                }
            }

            return compilationUnit;
        }

        protected Stream GetSystemResource() {
            return typeof(FixtureBase).Assembly.GetManifestResourceStream(typeof(FixtureBase).Namespace + ".System.trs");
        }

        private class ThrowingTraceListener : TraceListener {
            public override void Fail(string message, string detailMessage) {
                Assert.Fail(message);
            }

            public override void Write(string message) {
                Console.Write(message);
            }

            public override void WriteLine(string message) {
                Console.WriteLine(message);
            }
        }
    }
}
